docroot/modules/humsci/hs_field_helpers/hs_field_helpers.module

```diff
@@ -652,3 +653,70 @@ function hs_field_helpers_form_alter(&$form, FormStateInterface $form_state, $fo
     unset($form['header']['views_bulk_operations_bulk_form']['select_all']);
   }
 }
+
+/**
+ * Implements hook_field_widget_complete_WIDGET_TYPE_form_alter().
+ */
+function hs_field_helpers_field_widget_complete_link_default_form_alter(&$field_widget_complete_form, FormStateInterface $form_state, $context) {
+  // Add custom help text to link_default widgets.
+  if (!isset($field_widget_complete_form['widget']) || !is_array($field_widget_complete_form['widget'])) {
+    return;
+  }
+
+  $description = NULL;
+  $appended = NULL;
+
+  // Figure out if link supports internal, external, or both.
+  $field_definition = $context['items']->getFieldDefinition();
+  $settings = $field_definition->getSettings();
+  $link_type = $settings['link_type'];
+  $supports_internal = ($link_type & LinkItem::LINK_INTERNAL) === LinkItem::LINK_INTERNAL;
+  $supports_external = ($link_type & LinkItem::LINK_EXTERNAL) === LinkItem::LINK_EXTERNAL;
+
+  if (!$supports_internal && !$supports_external) {
+    return;
+  }
+
+  // Append extra text to the following form_ids.
+  $base_form = $form_state->getBuildInfo()['base_form_id'];
+  $append_extra_text_to = [
+    'shortcut_form',
+    'menu_link_content_form',
+  ];
+
+  if (in_array($base_form, $append_extra_text_to)) {
+    $appended = t('Enter %front to link to the front page. Enter %no_link to display a menu item that is only text without a link.', [
+      '%front' => '<front>',
+      '%no_link' => '<nolink>',
+    ]);
+  }
+
+  $placeholders = [
+    '%internal_link' => '/about/contact-us',
+    '%external_link' => 'https://www.stanford.edu',
+    '@appended' => $appended,
+  ];
+
+  // Internal link help text.
+  if ($supports_internal && !$supports_external) {
+    $description = t('This must be an internal path such as %internal_link You can also start typing the title of a piece of content to select it. @appended', $placeholders);
+  }
+
+  // External link help text.
+  if (!$supports_internal && $supports_external) {
+    $description = t('This must be an external URL such as %external_link. @appended', $placeholders);
+  }
+
+  // Both internal and external link help text.
+  if ($supports_internal && $supports_external) {
+    $description = t('Start typing the title of a piece of content to select it. You can also enter an internal path such as %internal_link or an external URL such as %external_link. @appended', $placeholders);
+  }
```

Comment:
All these cases are exclusive, so it's more performant to use `if - elseif - else` to prevent evaluating all conditions in all cases:

```
if ($supports_internal && !$supports_external) {
  // Internal only link.
  // ...
}
elseif (!$supports_internal && $supports_external) {
  // External only link.
  // ...
}
else {
  // Both internal and external link.
  // ...
}
```

Reply:
Good idea, fixed!